File: src/keboola_mcp_server/preview.py
import copy
import logging
from typing import Any

import jsonschema
import pydantic
import yaml
from pydantic import AliasChoices, BaseModel, Field, TypeAdapter, field_validator
from starlette.requests import Request
from starlette.responses import JSONResponse, Response

from keboola_mcp_server.clients import KeboolaClient
from keboola_mcp_server.clients.client import DATA_APP_COMPONENT_ID
from keboola_mcp_server.mcp import ServerState, SessionStateMiddleware
from keboola_mcp_server.tools import data_apps as data_app_tools
from keboola_mcp_server.tools.components import tools as components_tools
from keboola_mcp_server.tools.components.model import ConfigParamUpdate, TfParamUpdate
from keboola_mcp_server.tools.components.utils import get_sql_transformation_id_from_sql_dialect
from keboola_mcp_server.tools.flow import tools as flow_tools
from keboola_mcp_server.tools.flow.scheduler import list_schedules_for_config, validate_cron_tab
from keboola_mcp_server.tools.flow.scheduler_model import ScheduleDetail, ScheduleRequest
from keboola_mcp_server.tools.flow.tools import MODIFY_FLOW_TOOL_NAME, UPDATE_FLOW_TOOL_NAME
from keboola_mcp_server.workspace import WorkspaceManager

LOG = logging.getLogger(__name__)


class PreviewConfigDiffRq(BaseModel):
    tool_name: str = Field(
        validation_alias=AliasChoices('toolName', 'tool_name', 'tool-name', 'ToolName'),
        serialization_alias='toolName',
    )
    tool_params: dict[str, Any] = Field(
        validation_alias=AliasChoices('toolParams', 'tool_params', 'tool-params', 'ToolParams'),
        serialization_alias='toolParams',
    )


class ConfigCoordinates(BaseModel):
    component_id: str | None = Field(
        default=None,
        validation_alias=AliasChoices('componentId', 'component_id', 'component-id', 'ComponentId'),
        serialization_alias='componentId',
    )
    configuration_id: str | None = Field(
        default=None,
        validation_alias=AliasChoices('configurationId', 'configuration_id', 'configuration-id', 'ConfigurationId'),
        serialization_alias='configurationId',
    )
    configuration_row_id: str | None = Field(
        default=None,
        validation_alias=AliasChoices(
            'configurationRowId', 'configuration_row_id', 'configuration-row-id', 'ConfigurationRowId'
        ),
        serialization_alias='configurationRowId',
    )

    @field_validator('component_id', 'configuration_id', 'configuration_row_id', mode='before')
    @classmethod
    def convert_to_string(cls, v):
        """Convert any value to string, preserving None."""
        return None if v is None else str(v)


class PreviewConfigDiffResp(BaseModel):
    coordinates: ConfigCoordinates = Field(
        validation_alias=AliasChoices('coordinates', 'Coordinates'),
        serialization_alias='coordinates',
    )
    original_config: dict[str, Any] | None = Field(
        validation_alias=AliasChoices('originalConfig', 'original_config', 'original-config', 'OriginalConfig'),
        serialization_alias='originalConfig',
    )
    updated_config: dict[str, Any] | None = Field(
        validation_alias=AliasChoices('updatedConfig', 'updated_config', 'updated-config', 'UpdatedConfig'),
        serialization_alias='updatedConfig',
    )
    is_valid: bool = Field(
        validation_alias=AliasChoices('isValid', 'is_valid', 'is-valid', 'IsValid'),
        serialization_alias='isValid',
    )
    validation_errors: list[str] | None = Field(
        default=None,
        validation_alias=AliasChoices('validationErrors', 'validation_errors', 'validation-errors', 'ValidationErrors'),
        serialization_alias='validationErrors',
    )


async def _extract_coordinates(
    tool_name: str, tool_params: dict[str, Any], workspace_manager: WorkspaceManager
) -> ConfigCoordinates:
    """Extract configuration coordinates from tool parameters."""
    if tool_name == 'update_config':
        return ConfigCoordinates(
            component_id=tool_params.get('component_id'),
            configuration_id=tool_params.get('configuration_id'),
        )
    elif tool_name == 'update_config_row':
        return ConfigCoordinates(
            component_id=tool_params.get('component_id'),
            configuration_id=tool_params.get('configuration_id'),
            configuration_row_id=tool_params.get('configuration_row_id'),
        )
    elif tool_name == 'update_sql_transformation':
        sql_dialect = await workspace_manager.get_sql_dialect()
        return ConfigCoordinates(
            component_id=get_sql_transformation_id_from_sql_dialect(sql_dialect),
            configuration_id=tool_params.get('configuration_id'),
        )
    elif tool_name in {UPDATE_FLOW_TOOL_NAME, MODIFY_FLOW_TOOL_NAME}:
        return ConfigCoordinates(
            component_id=tool_params.get('flow_type'),
            configuration_id=tool_params.get('configuration_id'),
        )
    elif tool_name == 'modify_data_app':
        return ConfigCoordinates(
            component_id=DATA_APP_COMPONENT_ID,
            configuration_id=tool_params.get('configuration_id'),
        )
    else:
        raise ValueError(f'Invalid tool name: "{tool_name}"')


async def _validate_tool_params(
    tool_name: str,
    tool_params: dict[str, Any],
    tool_input_schema: dict[str, Any],
) -> tuple[bool, str | None]:
    """
    Validate tool parameters against the tool's input schema using JSON schema validation.

    This validates the parameters without executing the tool function.

    :param tool_name: Name of the MCP tool to validate against
    :param tool_params: Parameters to validate (raw user-provided params)
    :param tool_input_schema:
    :return: Tuple of (is_valid, validation_errors)
        - is_valid: True if validation passed, False otherwise
        - validation_errors: List of error messages if validation failed, None if successful
    """
    try:
        jsonschema.validate(instance=tool_params, schema=tool_input_schema)
        return True, None

    except jsonschema.ValidationError as e:
        # Format a validation error message similarly to errors.prettify_validation_error() function
        header = f'Found 1 validation error for {tool_name}:'
        formatted = {
            'errors': [
                {
                    'field': '.'.join(str(p) for p in e.path or []),
                    'message': e.message,
                    'extra': {
                        'schema': e.schema,
                    },
                }
            ]
        }
        try:
            yaml_str = yaml.dump(formatted, default_flow_style=False, sort_keys=False, allow_unicode=True)
        except Exception:
            yaml_str = str(formatted)

        return False, f'{header}\n{yaml_str}'

    except jsonschema.SchemaError as e:
        # Schema itself is invalid
        LOG.exception(f"[validate_tool_params] Invalid schema for tool '{tool_name}': {e}")
        return False, 'Internal error: Invalid tool schema'

    except Exception as e:
        # Handle other unexpected errors
        LOG.exception(f'[validate_tool_params] Unexpected error: {e}')
        return False, f'Validation error: {str(e)}'


def _prepare_mutator(
    preview_rq: PreviewConfigDiffRq, client: KeboolaClient, workspace_manager: WorkspaceManager
) -> tuple[Any, dict[str, Any]]:
    """
    Prepare mutator function and parameters for config diff preview.

    :param preview_rq: PreviewConfigDiffRq object containing tool parameters and name.
    :param client: KeboolaClient instance for API operations.
    :param workspace_manager: WorkspaceManager instance for workspace operations.
    :return: Tuple containing mutator function and parameters.
    """
    mutator_params: dict[str, Any] = {
        **preview_rq.tool_params,
        'client': client,
    }

    if preview_rq.tool_name == 'update_config':
        mutator_fn = components_tools.update_config_internal
        if parameter_updates := mutator_params.get('parameter_updates'):
            type_adapter = TypeAdapter(list[ConfigParamUpdate])
            mutator_params['parameter_updates'] = type_adapter.validate_python(parameter_updates)

    elif preview_rq.tool_name == 'update_config_row':
        mutator_fn = components_tools.update_config_row_internal
        if parameter_updates := mutator_params.get('parameter_updates'):
            type_adapter = TypeAdapter(list[ConfigParamUpdate])
            mutator_params['parameter_updates'] = type_adapter.validate_python(parameter_updates)

    elif preview_rq.tool_name == 'update_sql_transformation':
        mutator_fn = components_tools.update_sql_transformation_internal
        mutator_params['workspace_manager'] = workspace_manager
        if parameter_updates := mutator_params.get('parameter_updates'):
            type_adapter = TypeAdapter(list[TfParamUpdate])
            mutator_params['parameter_updates'] = type_adapter.validate_python(parameter_updates)

    elif preview_rq.tool_name in {UPDATE_FLOW_TOOL_NAME, MODIFY_FLOW_TOOL_NAME}:
        mutator_fn = flow_tools.update_flow_internal
        mutator_params.pop('schedules', None)

    elif preview_rq.tool_name == 'modify_data_app':
        mutator_fn = data_app_tools.modify_data_app_internal
        mutator_params['workspace_manager'] = workspace_manager

    else:
        raise ValueError(f'Invalid tool name: "{preview_rq.tool_name}"')

    return mutator_fn, mutator_params


def _apply_schedule_preview(
    schedules: list[ScheduleDetail],
    requests: list[ScheduleRequest],
) -> list[ScheduleDetail]:
    updated_schedules = [schedule.model_copy(deep=True) for schedule in schedules]
    schedule_index = {schedule.schedule_id: idx for idx, schedule in enumerate(updated_schedules)}
    new_index = 1

    for request in requests:
        validate_cron_tab(request.cron_tab)
        if request.action == 'add':
            if request.cron_tab is None:
                raise ValueError('Cron tab is required when creating a new schedule')
            schedule_id = f'preview-new-{new_index}'
            new_index += 1
            updated_schedules.append(
                ScheduleDetail(
                    schedule_id=schedule_id,
                    timezone=request.timezone or 'UTC',
                    state=request.state or 'enabled',
                    cron_tab=request.cron_tab,
                    target_executions=[],
                )
            )
            schedule_index[schedule_id] = len(updated_schedules) - 1
        elif request.action == 'update':
            if not request.schedule_id:
                raise ValueError('Schedule ID is required to update a schedule')
            schedule_idx = schedule_index.get(request.schedule_id)
            if schedule_idx is None:
                raise ValueError(f'Schedule ID "{request.schedule_id}" not found')
            schedule = updated_schedules[schedule_idx]
            updates: dict[str, Any] = {}
            if request.cron_tab is not None:
                updates['cron_tab'] = request.cron_tab
            if request.timezone is not None:
                updates['timezone'] = request.timezone
            if request.state is not None:
                updates['state'] = request.state
            if updates:
                updated_schedules[schedule_idx] = schedule.model_copy(update=updates)
        elif request.action == 'remove':
            if not request.schedule_id:
                raise ValueError('Schedule ID is required to remove a schedule')
            if request.schedule_id not in schedule_index:
                raise ValueError(f'Schedule ID "{request.schedule_id}" not found')
            updated_schedules = [schedule for schedule in updated_schedules if schedule.schedule_id != request.schedule_id]
            schedule_index = {schedule.schedule_id: idx for idx, schedule in enumerate(updated_schedules)}
        else:
            raise ValueError(f'Invalid action for schedulers: {request.action}.')

    return updated_schedules


async def preview_config_diff(rq: Request) -> Response:
    preview_rq = PreviewConfigDiffRq.model_validate(await rq.json())

    LOG.info(f'[preview_config_diff] {preview_rq}')
    LOG.info(f'[preview_config_diff] rq.app={rq.app}')
    LOG.info(f'[preview_config_diff] rq.app.state={rq.app.state} vars={vars(rq.app.state)}')
    LOG.info(f'[preview_config_diff] rq.state={rq.state} vars={vars(rq.state)}')

    server_state = ServerState.from_starlette(rq.app)
    config = SessionStateMiddleware.apply_request_config(rq, server_state.config)
    state = SessionStateMiddleware.create_session_state(config, server_state.runtime_info, readonly=True)
    client = KeboolaClient.from_state(state)
    workspace_manager = WorkspaceManager.from_state(state)

    coordinates = await _extract_coordinates(preview_rq.tool_name, preview_rq.tool_params, workspace_manager)

    if tool_input_schema := rq.app.state.mcp_tools_input_schema.get(preview_rq.tool_name):
        is_valid, validation_errors = await _validate_tool_params(
            tool_name=preview_rq.tool_name,
            tool_params=preview_rq.tool_params,
            tool_input_schema=tool_input_schema,
        )

        if not is_valid:
            preview_resp = PreviewConfigDiffResp(
                coordinates=coordinates,
                original_config=None,
                updated_config=None,
                is_valid=False,
                validation_errors=[validation_errors],
            )
            return JSONResponse(preview_resp.model_dump(by_alias=True, exclude_none=True))
    else:
        LOG.warning(f'[preview_config_diff] No input schema found for tool "{preview_rq.tool_name}"')

    mutator_fn, mutator_params = _prepare_mutator(preview_rq, client, workspace_manager)

    try:
        original_config, new_config, *_ = await mutator_fn(**mutator_params)
        if isinstance(original_config, BaseModel):
            original_config = original_config.model_dump()

        updated_config = copy.deepcopy(original_config)
        updated_config['configuration'] = new_config
        if name := preview_rq.tool_params.get('name'):
            updated_config['name'] = name
        description = preview_rq.tool_params.get('description') or preview_rq.tool_params.get('updated_description')
        if description:
            updated_config['description'] = description
        if change_description := preview_rq.tool_params.get('change_description'):
            updated_config['changeDescription'] = change_description
        if preview_rq.tool_name in {UPDATE_FLOW_TOOL_NAME, MODIFY_FLOW_TOOL_NAME}:
            schedule_payload = preview_rq.tool_params.get('schedules') or []
            if schedule_payload:
                type_adapter = TypeAdapter(list[ScheduleRequest])
                schedule_requests = type_adapter.validate_python(schedule_payload)
                flow_type = preview_rq.tool_params.get('flow_type')
                configuration_id = preview_rq.tool_params.get('configuration_id')
                if not flow_type or not configuration_id:
                    raise ValueError('Flow type and configuration ID are required to preview schedules')
                original_schedules = await list_schedules_for_config(
                    client=client,
                    component_id=flow_type,
                    configuration_id=configuration_id,
                )
                updated_schedules = _apply_schedule_preview(original_schedules, schedule_requests)
                original_config['schedules'] = [s.model_dump(by_alias=True) for s in original_schedules]
                updated_config['schedules'] = [s.model_dump(by_alias=True) for s in updated_schedules]

        preview_resp = PreviewConfigDiffResp(
            coordinates=coordinates,
            original_config=original_config,
            updated_config=updated_config,
            is_valid=True,
            validation_errors=None,
        )

    except (pydantic.ValidationError, jsonschema.ValidationError, ValueError) as ex:
        LOG.exception(f'[preview_config_diff] {ex}')
        preview_resp = PreviewConfigDiffResp(
            coordinates=coordinates,
            original_config=None,
            updated_config=None,
            is_valid=False,
            validation_errors=[str(ex)],
        )

    return JSONResponse(preview_resp.model_dump(by_alias=True, exclude_none=True))

File: src/preview.ts
/**
 * `/preview/configuration` custom HTTP endpoint — a faithful port of the Python
 * `preview.py` (`preview_config_diff`).
 *
 * It simulates a config-changing MCP tool ("update_config", "update_config_row", …)
 * WITHOUT writing anything, returning the original vs. updated configuration so a UI
 * can show a diff before the user commits. The route lives outside the MCP tool-call
 * path, so it re-applies the exact same authorization the real tool call would get:
 *
 * 1. Header authorization (`X-Allowed-Tools` / `X-Disallowed-Tools` / `X-Read-Only-Mode`)
 *    — port of `ToolAuthorizationMiddleware`.
 * 2. Project-feature / token-role / branch gating via {@link authorizeToolCall} — the
 *    AI-3438 hardening: without it a restricted caller could drive a write tool's
 *    preview (e.g. a data-app tool on a non-main branch, or any write tool with a
 *    read-only token).
 *
 * Everything runs against a READ-ONLY Keboola client: the raw clients are built with
 * `readonly: true`, so any accidental non-GET request throws instead of mutating.
 */

import type { ErrorObject } from 'ajv';
// zod emits draft 2020-12 JSON Schema, so use the matching Ajv build.
import Ajv2020 from 'ajv/dist/2020.js';
import { z } from 'zod';

import { createKeboolaClients, type KeboolaClients } from '@/clients/keboola';
import { createRawClient } from '@/clients/raw';
import { deriveServiceUrls } from '@/clients/urls';
import type { Config } from '@/config';
import { logger } from '@/logger';
import {
  hasAuthorizationFilters,
  isToolNameAuthorized,
  parseAuthorizationConfig,
} from '@/mcp/authorization';
import {
  authorizeToolCall,
  getProjectFeatures,
  getTokenRole,
  isSemanticToolName,
  type TokenInfo,
} from '@/mcp/filtering';
import { createServer } from '@/server';
import { configPreviewInternals } from '@/tools/components';
import type { ConfigParamUpdate } from '@/tools/components.model';

type JsonDict = Record<string, unknown>;

/** Error that carries an HTTP status code, mirroring the Python JSONResponse status codes. */
export class PreviewHttpError extends Error {
  constructor(
    message: string,
    readonly status: number,
  ) {
    super(message);
    this.name = 'PreviewHttpError';
  }
}

// ---------------------------------------------------------------------------
// Request / response shapes (port of the Pydantic models, including alias handling).
// ---------------------------------------------------------------------------

/** Reads the first present alias from a record, mirroring Pydantic `AliasChoices`. */
const pickAlias = (body: JsonDict, aliases: string[]): unknown => {
  for (const alias of aliases) {
    if (alias in body) return body[alias];
  }
  return undefined;
};

export type PreviewConfigDiffRq = {
  toolName: string;
  toolParams: JsonDict;
};

/** Parses the POST body into the request model (port of `PreviewConfigDiffRq.model_validate`). */
export const parsePreviewRequest = (raw: unknown): PreviewConfigDiffRq => {
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) {
    throw new PreviewHttpError('Request body must be a JSON object.', 400);
  }
  const body = raw as JsonDict;
  const toolName = pickAlias(body, ['toolName', 'tool_name', 'tool-name', 'ToolName']);
  const toolParams = pickAlias(body, ['toolParams', 'tool_params', 'tool-params', 'ToolParams']);
  if (typeof toolName !== 'string' || toolName.length === 0) {
    throw new PreviewHttpError('Field "toolName" is required and must be a string.', 400);
  }
  if (!toolParams || typeof toolParams !== 'object' || Array.isArray(toolParams)) {
    throw new PreviewHttpError('Field "toolParams" is required and must be an object.', 400);
  }
  return { toolName, toolParams: toolParams as JsonDict };
};

export type ConfigCoordinates = {
  componentId?: string | null;
  configurationId?: string | null;
  configurationRowId?: string | null;
};

export type PreviewConfigDiffResp = {
  coordinates: ConfigCoordinates;
  originalConfig: JsonDict | null;
  updatedConfig: JsonDict | null;
  isValid: boolean;
  validationErrors?: string[] | null;
};

const toStringOrNull = (value: unknown): string | null =>
  value === undefined || value === null ? null : String(value);

/**
 * Serializes the response with camelCase aliases and `exclude_none` semantics
 * (drops null/undefined entries, like Python's `model_dump(by_alias=True, exclude_none=True)`).
 */
const serializeResponse = (resp: PreviewConfigDiffResp): JsonDict => {
  const coordinates: JsonDict = {};
  if (resp.coordinates.componentId != null) coordinates.componentId = resp.coordinates.componentId;
  if (resp.coordinates.configurationId != null) {
    coordinates.configurationId = resp.coordinates.configurationId;
  }
  if (resp.coordinates.configurationRowId != null) {
    coordinates.configurationRowId = resp.coordinates.configurationRowId;
  }

  const out: JsonDict = { coordinates, isValid: resp.isValid };
  if (resp.originalConfig != null) out.originalConfig = resp.originalConfig;
  if (resp.updatedConfig != null) out.updatedConfig = resp.updatedConfig;
  if (resp.validationErrors != null) out.validationErrors = resp.validationErrors;
  return out;
};

// ---------------------------------------------------------------------------
// Tool metadata lookup (input schema + read-only hint), parity with the Python
// `app.state.mcp_tools_input_schema` / `mcp_read_only_tools` built from list_tools().
// ---------------------------------------------------------------------------

type RegisteredTool = {
  inputSchema?: z.ZodType;
  annotations?: { readOnlyHint?: boolean };
};

type ToolMetadata = {
  /** JSON Schema (draft) for the tool input, or undefined when the tool has no schema. */
  inputSchema?: JsonDict;
  isReadOnly: boolean;
};

/**
 * Enumerates the registered tools (unfiltered, like Python's `list_tools(run_middleware=False)`)
 * and returns the JSON Schema + read-only hint for the named tool. Building a server is the
 * single source of truth for tool schemas; no per-tool schema is duplicated here.
 */
const lookupToolMetadata = (config: Config, toolName: string): ToolMetadata | undefined => {
  const server = createServer(config) as unknown as {
    _registeredTools: Record<string, RegisteredTool>;
  };
  const tool = server._registeredTools[toolName];
  if (!tool) return undefined;
  // `io: 'input'` so fields with defaults stay optional (matching what the tool call
  // actually accepts, and the schema the MCP SDK advertises in tools/list).
  const inputSchema = tool.inputSchema
    ? (z.toJSONSchema(tool.inputSchema, { io: 'input' }) as JsonDict)
    : undefined;
  return { inputSchema, isReadOnly: tool.annotations?.readOnlyHint === true };
};

// ---------------------------------------------------------------------------
// Read-only client.
// ---------------------------------------------------------------------------

/**
 * Builds a Keboola client whose raw HTTP clients reject any non-GET request. The
 * preview path only reads (config detail, component fetch) + validates in memory, so a
 * read-only client is both sufficient and a safety net against accidental writes —
 * mirroring the Python `create_session_state(..., readonly=True)`.
 */
const createReadOnlyClients = (config: Config): KeboolaClients => {
  const clients = createKeboolaClients(config);
  if (!config.storageApiUrl) {
    throw new Error('Storage API URL is not configured.');
  }
  const urls = deriveServiceUrls(config.storageApiUrl);
  const token = config.storageToken!;
  const storageToken = config.bearerToken ? `Bearer ${config.bearerToken}` : token;
  return {
    ...clients,
    rawStorage: createRawClient({
      baseUrl: `${urls.storage}/v2/storage`,
      token: storageToken,
      readonly: true,
    }),
    rawQueue: createRawClient({ baseUrl: urls.queue, token, readonly: true }),
    rawAi: createRawClient({ baseUrl: urls.ai, token, readonly: true }),
    rawSyncActions: createRawClient({ baseUrl: urls.syncActions, token, readonly: true }),
  };
};

// ---------------------------------------------------------------------------
// Schema validation (port of `_validate_tool_params`, using ajv in place of jsonschema).
// ---------------------------------------------------------------------------

const formatValidationError = (toolName: string, error: ErrorObject): string => {
  // `instancePath` like "/parameter_updates/1" -> "parameter_updates.1" (parity with
  // Python's '.'.join(e.path)).
  const field = error.instancePath.replace(/^\//, '').replaceAll('/', '.');
  const message = error.message ?? 'is invalid';
  const detail = field ? `${field}: ${message}` : message;
  return `Found 1 validation error for ${toolName}:\n${detail}`;
};

/**
 * Validates raw tool params against the tool's JSON Schema. Returns the error message
 * (already formatted) when invalid, or null when valid. Mirrors `_validate_tool_params`.
 */
const validateToolParams = (
  toolName: string,
  toolParams: JsonDict,
  schema: JsonDict,
): string | null => {
  try {
    const ajv = new Ajv2020({ allErrors: false, strict: false });
    const validate = ajv.compile(schema);
    if (validate(toolParams)) return null;
    const first = validate.errors?.[0];
    if (!first) return `Found 1 validation error for ${toolName}:\nunknown validation error`;
    return formatValidationError(toolName, first);
  } catch (error) {
    logger.error({ err: error, tool: toolName }, '[preview] Invalid tool schema');
    return 'Internal error: Invalid tool schema';
  }
};

// ---------------------------------------------------------------------------
// Coordinate extraction (port of `_extract_coordinates`).
// ---------------------------------------------------------------------------

const UPDATE_FLOW_TOOL_NAME = 'update_flow';
const MODIFY_FLOW_TOOL_NAME = 'modify_flow';
const DATA_APP_COMPONENT_ID = 'keboola.data-apps';

const extractCoordinates = (toolName: string, toolParams: JsonDict): ConfigCoordinates => {
  switch (toolName) {
    case 'update_config':
      return {
        componentId: toStringOrNull(toolParams.component_id),
        configurationId: toStringOrNull(toolParams.configuration_id),
      };
    case 'update_config_row':
      return {
        componentId: toStringOrNull(toolParams.component_id),
        configurationId: toStringOrNull(toolParams.configuration_id),
        configurationRowId: toStringOrNull(toolParams.configuration_row_id),
      };
    case UPDATE_FLOW_TOOL_NAME:
    case MODIFY_FLOW_TOOL_NAME:
      return {
        componentId: toStringOrNull(toolParams.flow_type),
        configurationId: toStringOrNull(toolParams.configuration_id),
      };
    case 'modify_streamlit_data_app':
      return {
        componentId: DATA_APP_COMPONENT_ID,
        configurationId: toStringOrNull(toolParams.configuration_id),
      };
    case 'update_sql_transformation':
      // The Python endpoint resolves the component id from the workspace SQL dialect.
      // The diff for this tool is not implemented in the TS port (see DIFF_IMPLEMENTED_TOOLS),
      // so only the configuration id is surfaced here.
      return { configurationId: toStringOrNull(toolParams.configuration_id) };
    default:
      throw new PreviewHttpError(`Invalid tool name: "${toolName}"`, 400);
  }
};

// ---------------------------------------------------------------------------
// Mutator dispatch (port of `_prepare_mutator` + the diff assembly).
// ---------------------------------------------------------------------------

/** The set of config-changing tools the preview endpoint can diff. */
const SUPPORTED_TOOLS = new Set<string>([
  'update_config',
  'update_config_row',
  'update_sql_transformation',
  UPDATE_FLOW_TOOL_NAME,
  MODIFY_FLOW_TOOL_NAME,
  'modify_streamlit_data_app',
]);

/** Tools whose original/updated config diff is implemented in the TypeScript port. */
const DIFF_IMPLEMENTED_TOOLS = new Set<string>(['update_config', 'update_config_row']);

/**
 * Computes (originalConfig, newConfiguration) for the config tools whose mutation
 * internals are factored out and safe to run read-only. Returns the original config
 * record and the rebuilt `configuration` object.
 */
const computeConfigDiff = async (
  config: Config,
  clients: KeboolaClients,
  rq: PreviewConfigDiffRq,
): Promise<{ original: JsonDict; newConfiguration: JsonDict }> => {
  const params = rq.toolParams;
  const componentId = String(params.component_id ?? '');
  const configurationId = String(params.configuration_id ?? '');
  const parameterUpdates =
    (params.parameter_updates as ConfigParamUpdate[] | null | undefined) ?? null;
  const storage = (params.storage as JsonDict | null | undefined) ?? null;
  const processorsBefore = (params.processors_before as JsonDict[] | null | undefined) ?? null;
  const processorsAfter = (params.processors_after as JsonDict[] | null | undefined) ?? null;

  if (rq.toolName === 'update_config') {
    const original = await configPreviewInternals.configurationDetail(
      clients,
      componentId,
      configurationId,
    );
    const newConfiguration = await configPreviewInternals.buildUpdatedConfigPayload({
      config,
      clients,
      componentId,
      configurationId,
      parameterUpdates,
      storage,
      processorsBefore,
      processorsAfter,
      isRow: false,
    });
    return { original, newConfiguration };
  }

  // update_config_row
  const configurationRowId = String(params.configuration_row_id ?? '');
  const original = await configPreviewInternals.configurationRowDetail(
    clients,
    componentId,
    configurationId,
    configurationRowId,
  );
  const newConfiguration = await configPreviewInternals.buildUpdatedConfigPayload({
    config,
    clients,
    componentId,
    configurationId,
    configurationRowId,
    parameterUpdates,
    storage,
    processorsBefore,
    processorsAfter,
    isRow: true,
  });
  return { original, newConfiguration };
};

/**
 * Assembles the updated config from the original + new `configuration` and the
 * top-level field overrides (name/description/changeDescription/isDisabled), mirroring
 * the diff assembly block in `preview_config_diff`.
 */
const assembleUpdatedConfig = (
  original: JsonDict,
  newConfiguration: JsonDict,
  toolParams: JsonDict,
): JsonDict => {
  const updated = structuredClone(original);
  updated.configuration = newConfiguration;
  if (toolParams.name) updated.name = toolParams.name;
  if (toolParams.description) updated.description = toolParams.description;
  if (toolParams.is_disabled !== undefined && toolParams.is_disabled !== null) {
    updated.isDisabled = toolParams.is_disabled;
  }
  if (toolParams.change_description) updated.changeDescription = toolParams.change_description;
  return updated;
};

// ---------------------------------------------------------------------------
// Main handler.
// ---------------------------------------------------------------------------

/**
 * Verifies the Storage token to read project features + admin role. Throwing here would
 * surface as a 500; on failure we degrade to no-feature / no-role defaults so gating is
 * still applied (and a misconfigured token simply fails the project gate).
 */
const verifyToken = async (clients: KeboolaClients): Promise<TokenInfo> => {
  try {
    return (await clients.storage.tokens.verify()) as TokenInfo;
  } catch {
    return {};
  }
};

/**
 * Runs the full preview flow for an already-parsed request and returns the response
 * payload, throwing {@link PreviewHttpError} for the 4xx cases. The Hono route is a thin
 * wrapper around this.
 */
export const runPreviewConfigDiff = async (
  config: Config,
  rq: PreviewConfigDiffRq,
): Promise<JsonDict> => {
  const meta = lookupToolMetadata(config, rq.toolName);
  const isReadOnly = meta?.isReadOnly ?? false;

  // 1) Header authorization (port of ToolAuthorizationMiddleware). This route runs
  // outside the MCP middleware chain, so enforce the same headers explicitly.
  const auth = parseAuthorizationConfig({
    allowedTools: config.allowedTools,
    disallowedTools: config.disallowedTools,
    readOnlyMode: config.readOnlyMode,
  });
  if (hasAuthorizationFilters(auth) && !isToolNameAuthorized(rq.toolName, isReadOnly, auth)) {
    logger.info(`[preview] Tool authorization denied (headers): ${rq.toolName}`);
    throw new PreviewHttpError(`The tool "${rq.toolName}" is not authorized for this client.`, 403);
  }

  // Log only non-sensitive metadata; toolParams can carry user-supplied secrets.
  logger.info(
    `[preview] toolName=${rq.toolName} paramKeys=${Object.keys(rq.toolParams).sort().join(',')}`,
  );

  const clients = createReadOnlyClients(config);

  // 2) Project-feature / token-role / branch gating (the AI-3438 hardening) via the
  // exact same authorizeToolCall decision a real MCP tool call uses.
  const tokenInfo = await verifyToken(clients);
  const denial = authorizeToolCall({
    toolName: rq.toolName,
    isReadOnly,
    isSemantic: isSemanticToolName(rq.toolName),
    tokenRole: getTokenRole(tokenInfo),
    features: getProjectFeatures(tokenInfo),
    isOauth: Boolean(config.bearerToken),
    isMainBranch: config.branchId === undefined,
  });
  if (denial) {
    logger.info(`[preview] Tool authorization denied (project/role/branch): ${rq.toolName}`);
    throw new PreviewHttpError(denial, 403);
  }

  const coordinates = extractCoordinates(rq.toolName, rq.toolParams);

  // 3) Validate the params against the tool's input schema. A schema failure returns a
  // 200 with isValid=false (empty configs) — the KAI backend relies on that shape.
  if (meta?.inputSchema) {
    const validationError = validateToolParams(rq.toolName, rq.toolParams, meta.inputSchema);
    if (validationError) {
      return serializeResponse({
        coordinates,
        originalConfig: {},
        updatedConfig: {},
        isValid: false,
        validationErrors: [validationError],
      });
    }
  } else {
    logger.warn(`[preview] No input schema found for tool "${rq.toolName}"`);
  }

  if (!SUPPORTED_TOOLS.has(rq.toolName)) {
    throw new PreviewHttpError(`Invalid tool name: "${rq.toolName}"`, 400);
  }

  // The mutation internals for update_sql_transformation / update_flow / modify_flow /
  // modify_streamlit_data_app are not factored out as reusable, write-free functions in
  // the TypeScript port (they live inline in the tool handlers and perform the write),
  // so their read-only diff is not yet available. Authorization, validation and
  // coordinate extraction above still run for them.
  if (!DIFF_IMPLEMENTED_TOOLS.has(rq.toolName)) {
    throw new PreviewHttpError(
      `Configuration diff preview for tool "${rq.toolName}" is not supported by this server.`,
      400,
    );
  }

  // 4) Run the (read-only) mutator and assemble the diff. A value/validation error
  // returns a 200 with isValid=false (parity with the Python except block).
  try {
    const { original, newConfiguration } = await computeConfigDiff(config, clients, rq);
    const updatedConfig = assembleUpdatedConfig(original, newConfiguration, rq.toolParams);
    return serializeResponse({
      coordinates,
      originalConfig: original,
      updatedConfig,
      isValid: true,
      validationErrors: null,
    });
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    logger.info(`[preview] ${message}`);
    return serializeResponse({
      coordinates,
      originalConfig: {},
      updatedConfig: {},
      isValid: false,
      validationErrors: [message],
    });
  }
};
